vopono_core/src/network/piapf.rs

```diff
@@ -0,0 +1,193 @@
+            Protocol::Wireguard => pia.load_wireguard_auth()?,
+            _ => {
+                log::error!("PIA port forwarding only supported for OpenVPN and Wireguard");
+                anyhow::bail!("PIA port forwarding only supported for OpenVPN and Wireguard")
+            }
+        };
+        
+        //log::info!("PIA u/p: {} / {}", pia_user, pia_pass);
+        
+        let pia_token = PrivateInternetAccess::get_pia_token(&pia_user, &pia_pass)?;
+        
+        log::info!("PIA pia_token: {}", pia_token);
+
+        let get_response = NetworkNamespace::exec_with_output(&ns.name, &["curl", 
+            "-s", "-m", "5",
+            "--connect-to", &format!("{}::{}:", vpn_hostname, vpn_gateway).to_string(),
+            "--cacert", "/home/benland100/vopono/vopono_core/src/config/providers/pia/ca.rsa.4096.crt", //FIXME: how to get this path?
+            "-G", "--data-urlencode", &format!("token={}",pia_token).to_string(),
+            &format!("https://{}:19999/getSignature",vpn_hostname).to_string() ] )?;
+        if !get_response.status.success() {
+            log::error!("Could not obtain signature for port forward from PIA API");
+            anyhow::bail!("Could not obtain signature for port forward from PIA API")
+        }
+        
+        let parsed = json::parse(String::from_utf8_lossy(&get_response.stdout).as_ref())?;
+        if parsed["status"] != "OK" {
+            log::error!("Signature for port forward from PIA API not OK");
+            anyhow::bail!("Signature for port forward from PIA API not OK");
+        }
+        
+        let signature = parsed["signature"].as_str().unwrap().to_string();
+        let payload = parsed["payload"].as_str().unwrap().to_string();
+        let decoded = BASE64_STANDARD.decode(&payload)?;
+        let parsed = json::parse(String::from_utf8_lossy(&decoded).as_ref())?;
+        let port = parsed["port"].as_u16().unwrap();
+
+        let params = ThreadParams {
+            netns_name: ns.name.clone(),
+            hostname: vpn_hostname,
+            gateway: vpn_gateway,
+            signature: signature,
+            payload: payload,
+            port: port,
+        };
+        Self::refresh_port(&params)?;
+        let (send, recv) = mpsc::channel::<bool>();
+        let handle = std::thread::spawn(move || Self::thread_loop(params, recv));
+
+        log::info!("PIA forwarded local port: {port}");
+        Ok(Self {
+            port: port,
+            loop_thread_handle: Some(handle),
+            send_channel: send,
+        })
+    }
+
+    fn refresh_port(params: &ThreadParams) -> anyhow::Result<u16> {
+    
+        let bind_response = NetworkNamespace::exec_with_output(&params.netns_name, &["curl", 
+            "-Gs", "-m", "5",
+            "--connect-to", &format!("{}::{}:", params.hostname, params.gateway).to_string(),
+            "--cacert", "/home/benland100/vopono/vopono_core/src/config/providers/pia/ca.rsa.4096.crt", //FIXME: how to get this path?
```

Comment:
As above, we have to download it at sync time if possible.

Reply:
Note we actually already embed it in the binary - https://github.com/jamesmcm/vopono/blob/master/vopono_core/src/config/providers/pia/wireguard.rs#L94 - so we could just write that out during the sync process (or download it and avoid embedding it, but I assume it wasn't publicly accessible when I chose to embed it).